File: framework/languages/dotnet/e2e/PubSub/Client/Support/NetworkFaultProxy.cs
using System.Net;
using System.Net.Sockets;

namespace PubSub.Client.Support;

internal sealed class NetworkFaultProxy : IAsyncDisposable
{
    private readonly CancellationTokenSource _stop = new();
    private readonly TcpListener _listener = new(IPAddress.Loopback, 0);
    private readonly Uri _upstream;
    private readonly SemaphoreSlim _networkGate = new(1, 1);
    private readonly SemaphoreSlim _connected = new(0);
    private readonly object _connectionGate = new();
    private readonly Task _run;
    private TcpClient? _client;
    private TcpClient? _server;
    private bool _blocked;
    private long _clientToServerBytes;
    private long _serverToClientBytes;
    private string? _lastPumpFailure;

    public NetworkFaultProxy(Uri upstream, bool initiallyBlocked = false)
    {
        _upstream = upstream;
        if (initiallyBlocked)
        {
            _networkGate.Wait();
            _blocked = true;
        }
        _listener.Start();
        Endpoint = new Uri($"tcp://127.0.0.1:{((IPEndPoint)_listener.LocalEndpoint).Port}");
        _run = RunAsync(_stop.Token);
    }

    public Uri Endpoint { get; }

    public string Diagnostics =>
        $"clientToServer={Interlocked.Read(ref _clientToServerBytes)} "
        + $"serverToClient={Interlocked.Read(ref _serverToClientBytes)} "
        + $"lastFailure={_lastPumpFailure ?? "<none>"}";

    public void Block()
    {
        lock (_connectionGate)
        {
            if (_blocked) return;
            if (!_networkGate.Wait(0))
                throw new InvalidOperationException("The network fault proxy gate is already held.");
            _blocked = true;
            _client?.Dispose();
            _server?.Dispose();
        }
    }

    public void Unblock()
    {
        lock (_connectionGate)
        {
            if (!_blocked) return;
            _blocked = false;
            _networkGate.Release();
        }
    }

    public async Task WaitForUpstreamConnectionAsync(CancellationToken cancellationToken = default)
    {
        if (!await _connected.WaitAsync(TimeSpan.FromSeconds(15), cancellationToken))
            throw new TimeoutException("The network fault proxy did not connect to its upstream endpoint.");
    }

    private async Task RunAsync(CancellationToken cancellationToken)
    {
        while (!cancellationToken.IsCancellationRequested)
        {
            using var client = await _listener.AcceptTcpClientAsync(cancellationToken);
            client.NoDelay = true;
            await _networkGate.WaitAsync(cancellationToken);
            _networkGate.Release();
            using var server = new TcpClient();
            server.NoDelay = true;
            await server.ConnectAsync(_upstream.Host, _upstream.Port, cancellationToken);
            lock (_connectionGate)
            {
                _client = client;
                _server = server;
            }
            _connected.Release();
            try
            {
                await Task.WhenAny(
                    PumpAsync(client.GetStream(), server.GetStream(), true, cancellationToken),
                    PumpAsync(server.GetStream(), client.GetStream(), false, cancellationToken));
            }
            catch (Exception error) when (
                error is IOException or ObjectDisposedException or SocketException)
            {
                _lastPumpFailure = $"{error.GetType().Name}:{error.Message}";
            }
            finally
            {
                lock (_connectionGate)
                {
                    _client = null;
                    _server = null;
                }
            }
        }
    }

    private async Task PumpAsync(
        NetworkStream source,
        NetworkStream target,
        bool clientToServer,
        CancellationToken cancellationToken)
    {
        var buffer = new byte[16 * 1024];
        while (true)
        {
            var count = await source.ReadAsync(buffer, cancellationToken);
            if (count == 0) return;
            await target.WriteAsync(buffer.AsMemory(0, count), cancellationToken);
            if (clientToServer)
                Interlocked.Add(ref _clientToServerBytes, count);
            else
                Interlocked.Add(ref _serverToClientBytes, count);
        }
    }

    public async ValueTask DisposeAsync()
    {
        Unblock();
        _stop.Cancel();
        _listener.Stop();
        lock (_connectionGate)
        {
            _client?.Dispose();
            _server?.Dispose();
        }
        try
        {
            await _run;
        }
        catch (Exception error) when (
            error is OperationCanceledException or ObjectDisposedException or SocketException)
        {
        }
        _connected.Dispose();
        _networkGate.Dispose();
        _stop.Dispose();
    }
}

File: framework/languages/node/e2e/PubSub/Client/Support/network-fault-proxy.ts
import net from 'node:net';
import type { Server, Socket } from 'node:net';

export class NetworkFaultProxy {
  private readonly sockets = new Set<Socket>();

  private constructor(
    private readonly server: Server,
    private readonly upstream: URL,
    private blocked: boolean,
    readonly endpoint: string,
    readonly redisEndpoint: string,
    readonly port: number
  ) {
    server.on('connection', (client) => this.accept(client));
  }

  static async start(
    upstreamEndpoint: string,
    initiallyBlocked = false,
    listenPort = 0
  ): Promise<NetworkFaultProxy> {
    const server = net.createServer();
    await new Promise<void>((resolve, reject) => {
      server.once('error', reject);
      server.listen(listenPort, '127.0.0.1', resolve);
    });
    const address = server.address();
    if (typeof address !== 'object' || address === null) throw new Error('Network fault proxy did not bind a TCP port.');
    const normalizedUpstream = upstreamEndpoint.includes('://')
      ? upstreamEndpoint
      : `tcp://${upstreamEndpoint}`;
    return new NetworkFaultProxy(
      server,
      new URL(normalizedUpstream.replace(/^tcp:/, 'http:')),
      initiallyBlocked,
      `tcp://127.0.0.1:${address.port}`,
      `127.0.0.1:${address.port}`,
      address.port
    );
  }

  block(): void {
    this.blocked = true;
    for (const socket of this.sockets) socket.destroy();
    this.sockets.clear();
  }

  unblock(): void {
    this.blocked = false;
  }

  async close(): Promise<void> {
    this.block();
    await new Promise<void>((resolve, reject) => this.server.close((error) => error === undefined ? resolve() : reject(error)));
  }

  private accept(client: Socket): void {
    if (this.blocked) {
      client.destroy();
      return;
    }
    const upstream = net.connect(Number(this.upstream.port), this.upstream.hostname);
    this.track(client);
    this.track(upstream);
    upstream.once('connect', () => {
      client.pipe(upstream);
      upstream.pipe(client);
    });
    upstream.once('error', () => client.destroy());
    client.once('error', () => upstream.destroy());
  }

  private track(socket: Socket): void {
    this.sockets.add(socket);
    socket.once('close', () => this.sockets.delete(socket));
  }
}
